File: app/src/main/java/org/ymkm/pwchart/TopAct.java
package org.ymkm.pwchart;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.util.ArrayMap;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayout;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.PopupWindow;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import passwordchart.ymkm.org.passwordchart.R;

public class TopAct extends AppCompatActivity {

    @BindView(R.id.tv_seed)
    EditText tvSeed;
    @BindView(R.id.tv_pw)
    EditText tvPw;
    @BindView(R.id.use_numbers)
    Switch useNumbers;
    @BindView(R.id.use_symbols)
    Switch useSymbols;
    @BindView(R.id.tv_generated_password)
    TextView tvGeneratedPassword;
    @BindView(R.id.generated_password)
    LinearLayout generatedPassword;
    @BindView(R.id.tv_chart_number)
    TextView tvChartNumber;
    @BindView(R.id.char_table)
    GridLayout charTable;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_top);
        ButterKnife.bind(this);
        setupViews();
    }


    @OnCheckedChanged({R.id.use_numbers, R.id.use_symbols})
    public void onChartTypeChanged(@NonNull final CompoundButton btn, final boolean isChecked) {
        switch (btn.getId()) {
            case R.id.use_numbers:
                mChartType = isChecked ? (mChartType | PWChartUtils.CHART_TYPE_INCLUDE_NUMBERS)
                                     : (mChartType & ~PWChartUtils.CHART_TYPE_INCLUDE_NUMBERS);
                break;
            case R.id.use_symbols:
                mChartType = isChecked ? (mChartType | PWChartUtils.CHART_TYPE_INCLUDE_SYMBOLS)
                                     : (mChartType & ~PWChartUtils.CHART_TYPE_INCLUDE_SYMBOLS);
                break;
        }
        updateChart();
        updatePassword();
    }

    @OnTextChanged({R.id.tv_pw, R.id.tv_seed})
    public void onInputChanged() {
        updateChart();
        updatePassword();
    }

    @OnClick(R.id.tv_generated_password)
    public void onPwClicked() {
        final String pw = tvGeneratedPassword.getText().toString();
        ClipboardManager clipboard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText(pw, pw);
        clipboard.setPrimaryClip(clip);
        Toast.makeText(this, R.string.toast_copied_pw_to_clipboard_text, Toast.LENGTH_SHORT).show();
    }


    private void updateChart() {
        tvChartNumber.setVisibility(TextUtils.isEmpty(tvSeed.getText())
                                            ? View.INVISIBLE
                                            : View.VISIBLE);
        charTable.setVisibility(TextUtils.isEmpty(tvSeed.getText())
                                        ? View.INVISIBLE
                                        : View.VISIBLE);

        if (TextUtils.isEmpty(tvSeed.getText())) {
            return;
        }
        final String seedStr = tvSeed.getText().toString();
        final long chartNum = PWChartUtils.chartNumber(seedStr, mChartType);
        tvChartNumber.setText(TextUtils.expandTemplate(getString(R.string.pwchart_number), String.valueOf(chartNum)));
        updateCharTable(seedStr, useNumbers.isChecked(), useSymbols.isChecked());
    }

    private void updatePassword() {
        generatedPassword.setVisibility(TextUtils.isEmpty(tvSeed.getText()) || TextUtils.isEmpty(tvPw.getText())
                                                ? View.INVISIBLE
                                                : View.VISIBLE);
        if (TextUtils.isEmpty(tvPw.getText()) || TextUtils.isEmpty(tvSeed.getText())) {
            return;
        }

        final String seedStr = tvSeed.getText().toString();
        final String pw = tvPw.getText().toString().toUpperCase(Locale.getDefault());
        final String genPw = PWChartUtils.genPasswordWithSeed(seedStr, pw, useNumbers.isChecked(), useSymbols.isChecked());
        tvGeneratedPassword.setText(genPw);
    }

    private void updateCharTable(@NonNull final String seed, final boolean useNumbers, final boolean useSymbols) {
        charTable.setVisibility(TextUtils.isEmpty(tvSeed.getText())
                                        ? View.INVISIBLE
                                        : View.VISIBLE);
        final ArrayMap<Character, String> map = PWChartUtils.getTable(seed, useNumbers, useSymbols);
        for (int i = 0; i < charTable.getChildCount(); ++i) {
            final Button b = (Button) charTable.getChildAt(i);
            final char charMap = b.getText().toString().charAt(0);
            b.setOnClickListener(new View.OnClickListener() {

                @Override
                public void onClick(final View v) {
                    if (charTable.getTag() != null) {
                        ((View) charTable.getTag()).setActivated(false);
                    }
                    if (mShownPopup != null) {
                        ((TextView) mShownPopup.getContentView()).setText(map.get(charMap));
                    }
                    else {
                        final TextView tv = (TextView) LayoutInflater.from(TopAct.this).inflate(R.layout.char_popup, null);
                        tv.setText(map.get(charMap));
                        tv.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED);
                        mShownPopup = new PopupWindow(tv, tv.getMeasuredWidth(), tv.getMeasuredHeight());
                        charTable.setTag(b);
                    }
                    mShownPopup.setOutsideTouchable(true);
                    mShownPopup.showAtLocation(v, Gravity.CENTER, 0, 0);
                    mShownPopup.setOnDismissListener(new PopupWindow.OnDismissListener() {

                        @Override
                        public void onDismiss() {
                            if (charTable.getTag() != null) {
                                ((View) charTable.getTag()).setActivated(false);
                            }
                        }
                    });
                    charTable.setTag(b);
                    b.setActivated(true);
                }
            });
        }
    }


    private void setupViews() {
        updateChart();
        updatePassword();
    }


    private PopupWindow mShownPopup;
    private int mChartType = PWChartUtils.CHART_TYPE_NORMAL;
}

File: app/src/main/java/org/ymkm/pwchart/TopAct.kt
package org.ymkm.pwchart

import android.annotation.SuppressLint
import android.content.ClipData
import android.content.ClipboardManager
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.PopupWindow
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.core.view.children
import androidx.core.view.isInvisible
import androidx.databinding.BindingAdapter
import androidx.databinding.DataBindingUtil
import androidx.databinding.ObservableBoolean
import androidx.databinding.ObservableField
import org.ymkm.passwordchart.R
import org.ymkm.passwordchart.databinding.ActivityTopBinding
import java.util.Locale

class TopAct : AppCompatActivity(), ViewCallback {
    private lateinit var binding: ActivityTopBinding

    override val limitChars: ObservableBoolean = ObservableBoolean(false)
    override fun limitChars(limit: Boolean) {
        limitChars.set(limit)
    }

    override var includeNumbers: Boolean = true
        set(isChecked) {
            field = isChecked
            if (isChecked) chartType or PWChartUtils.CHART_TYPE_INCLUDE_NUMBERS else chartType and PWChartUtils.CHART_TYPE_INCLUDE_NUMBERS.inv()
            updateChart()
            updatePassword()
        }
    override var includeSymbols: Boolean = true
        set(isChecked) {
            field = isChecked
            if (isChecked) chartType or PWChartUtils.CHART_TYPE_INCLUDE_SYMBOLS else chartType and PWChartUtils.CHART_TYPE_INCLUDE_SYMBOLS.inv()
            updateChart()
            updatePassword()
        }
    override val chartNumber: ObservableField<String?> = ObservableField()
    override val generatedPassword: ObservableField<String?> = ObservableField()

    override fun onCreate(savedInstanceState: Bundle?) {
        installSplashScreen().setKeepOnScreenCondition { false }
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_top)
        binding.callback = this
    }

    override fun onInputChanged() {
        updateChart()
        updatePassword()
    }

    override fun onPwClicked() {
        val generatedPassword = generatedPassword.get() ?: return
        val clipboard = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText(generatedPassword, generatedPassword)
        clipboard.setPrimaryClip(clip)
        Toast.makeText(this, R.string.toast_copied_pw_to_clipboard_text, Toast.LENGTH_SHORT).show()
    }

    private fun updateChart() {
        val seedStr = binding.tvSeed.text.toString()
        binding.tvChartNumber.isInvisible = seedStr.isEmpty()
        binding.charTable.isInvisible = seedStr.isEmpty()
        if (seedStr.isEmpty()) {
            chartNumber.set(null)
            return
        }
        val chartNum = PWChartUtils.chartNumber(seedStr, chartType)
        chartNumber.set(
            TextUtils.expandTemplate(
                getString(R.string.pwchart_number),
                chartNum.toString()
            ).toString()
        )
        updateCharTable(
            seed = seedStr,
            useNumbers = binding.useNumbers.isChecked,
            useSymbols = binding.useSymbols.isChecked
        )
    }

    private fun updatePassword() {
        val seedStr: String = binding.tvSeed.getText().toString()
        val pw: String = binding.tvPw.text.toString().uppercase(Locale.getDefault())
        if (seedStr.isEmpty() || pw.isEmpty()) {
            _generatedPassword = null
            updateGeneratedPassword()
            return
        }
        _generatedPassword =
            PWChartUtils.genPasswordWithSeed(
                seedStr,
                pw,
                includeNumbers,
                includeSymbols
            )
        updateGeneratedPassword()
    }

    private fun updateGeneratedPassword() {
        val currentPassword = _generatedPassword.orEmpty()
        val passMaxLength = maxChars ?: currentPassword.length
        generatedPassword.set(currentPassword.take(passMaxLength))
    }

    @SuppressLint("InflateParams")
    private fun updateCharTable(seed: String, useNumbers: Boolean, useSymbols: Boolean) {
        binding.charTable.isInvisible = binding.tvSeed.text.isEmpty()
        val map = PWChartUtils.getTable(seed, useNumbers, useSymbols)
        binding.charTable.children.forEach { child ->
            val b = child as Button
            val charMap = b.text[0]
            b.setOnClickListener { v ->
                if (binding.charTable.tag != null) {
                    (binding.charTable.tag as View).isActivated = false
                }
                val currentPopup = if (shownPopup == null) {
                    val tv = LayoutInflater.from(this@TopAct)
                        .inflate(R.layout.char_popup, null) as TextView
                    tv.text = map[charMap]
                    tv.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED)
                    shownPopup = PopupWindow(tv, tv.measuredWidth, tv.measuredHeight)
                    binding.charTable.tag = b
                    requireNotNull(shownPopup)
                } else {
                    requireNotNull(shownPopup).also {
                        (it.contentView as TextView).text = map[charMap]
                    }
                }
                currentPopup.isOutsideTouchable = true
                currentPopup.showAtLocation(v, Gravity.CENTER, 0, 0)
                currentPopup.setOnDismissListener {
                    if (binding.charTable.tag != null) {
                        (binding.charTable.tag as View).isActivated = false
                    }
                }
                binding.charTable.tag = b
                b.isActivated = true
            }
        }
    }

    override fun maxChars(editable: Editable?) {
        maxChars = editable?.toString()?.toIntOrNull()
        updateGeneratedPassword()
    }

    private var shownPopup: PopupWindow? = null
    private var chartType = PWChartUtils.CHART_TYPE_NORMAL
    private var maxChars: Int? = null
    private var _generatedPassword: String? = null
}

@BindingAdapter("enabled")
fun TextView.enabled(isEnabled: Boolean) {
    this.isEnabled = isEnabled
}

@BindingAdapter("isVisible")
fun View.isVisible(isVisible: Boolean?) {
    this.visibility = if (isVisible == true) View.VISIBLE else View.GONE
}

@BindingAdapter("isInvisible")
fun View.isInvisible(isInvisible: Boolean?) {
    this.visibility = if (isInvisible == true) View.INVISIBLE else View.VISIBLE
}